FP16 precision causing MultiLayerNetwork.fit() crash Android

####Issue Description
Trying to use FP16 precision in Android causes crashes in the MultiLayerNetwork.fit() method, while using the standard FLOAT precision, no problems are encountered. For example:

    Nd4j.setDataType(DataType.HALF);
    DataSetIterator dataIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, outputNum);
    DataNormalization scalar = new ImagePreProcessingScaler(0, 1);
    scalar.fit(dataIter);
    dataIter.setPreProcessor(scalar);
    network.fit(dataIter);
the MultiLayerNetwork.fit() method crashes


####Version Information
Deeplearning4j version 1.0.0-SNAPSHOT in Android

####Stack trace

E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
    Process: com.example.robrecht.mnistest, PID: 6019
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:365)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:784)
     Caused by: java.lang.UnsupportedOperationException
        at org.nd4j.linalg.cpu.nativecpu.blas.CpuLevel3.hgemm(CpuLevel3.java:44)
        at org.nd4j.linalg.api.blas.impl.BaseLevel3.gemm(BaseLevel3.java:74)
        at org.nd4j.linalg.api.ndarray.BaseNDArray.mmuli(BaseNDArray.java:3727)
        at org.deeplearning4j.nn.layers.BaseLayer.preOutputWithPreNorm(BaseLayer.java:318)
        at org.deeplearning4j.nn.layers.BaseLayer.preOutput(BaseLayer.java:293)
        at org.deeplearning4j.nn.layers.BaseLayer.activate(BaseLayer.java:339)
        at org.deeplearning4j.nn.layers.AbstractLayer.activate(AbstractLayer.java:259)
        at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.ffToLayerActivationsInWs(MultiLayerNetwork.java:1099)
        at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2622)
        at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2580)
        at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:160)
        at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
        at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
        at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fitHelper(MultiLayerNetwork.java:1653)
        at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1573)
        at com.example.robrecht.mnistest.MNISTTester.trainNetwork(MNISTTester.java:213)
        at com.example.robrecht.mnistest.MNISTTester$AsyncTaskTrainer.doInBackground(MNISTTester.java:303)
        at com.example.robrecht.mnistest.MNISTTester$AsyncTaskTrainer.doInBackground(MNISTTester.java:293)
        at android.os.AsyncTask$2.call(AsyncTask.java:345)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:784)

--- a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/blas/impl/BaseLevel2.java
+++ b/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/blas/impl/BaseLevel2.java
@@ -16,6 +16,7 @@
 
 package org.nd4j.linalg.api.blas.impl;
 
+import lombok.val;
 import org.nd4j.linalg.api.blas.Level2;
 import org.nd4j.linalg.api.blas.params.GemvParameters;
 import org.nd4j.linalg.api.buffer.DataBuffer;
@@ -24,6 +25,7 @@ import org.nd4j.linalg.api.ndarray.INDArray;
 import org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner;
 import org.nd4j.linalg.api.ops.executioner.OpExecutioner;
 import org.nd4j.linalg.api.ops.executioner.OpExecutionerUtil;
+import org.nd4j.linalg.exception.ND4JIllegalStateException;
 import org.nd4j.linalg.factory.Nd4j;
 import org.nd4j.linalg.profiler.OpProfiler;
 
@@ -66,12 +68,28 @@ public abstract class BaseLevel2 extends BaseLevel implements Level2 {
             dgemv(order, parameters.getAOrdering(), parameters.getM(), parameters.getN(), alpha, parameters.getA(),
                             parameters.getLda(), parameters.getX(), parameters.getIncx(), beta, parameters.getY(),
                             parameters.getIncy());
-        } else {
+        } else if (A.data().dataType() == DataType.FLOAT){
             DefaultOpExecutioner.validateDataType(DataType.FLOAT, parameters.getA(), parameters.getX(),
                             parameters.getY());
             sgemv(order, parameters.getAOrdering(), parameters.getM(), parameters.getN(), (float) alpha,
                             parameters.getA(), parameters.getLda(), parameters.getX(), parameters.getIncx(),
                             (float) beta, parameters.getY(), parameters.getIncy());
+        } else if (A.data().dataType() == DataType.HALF) {
+            DefaultOpExecutioner.validateDataType(DataType.HALF, parameters.getA(), parameters.getX(),
+                    parameters.getY());
+
+            // TODO: provide optimized GEMV kernel eventually
+            val fA = parameters.getA().castTo(DataType.FLOAT);
+            val fX = parameters.getX().castTo(DataType.FLOAT);
+            val fY = parameters.getY().castTo(DataType.FLOAT);
+
+            sgemv(order, parameters.getAOrdering(), parameters.getM(), parameters.getN(), (float) alpha,
+                    fA, parameters.getLda(), fX, parameters.getIncx(),
+                    (float) beta, fY, parameters.getIncy());
+
+            Y.assign(fY);
+        } else {
+            throw new ND4JIllegalStateException("Unsupported data type " + A.dataType());
         }
 
         OpExecutionerUtil.checkForAny(Y);

--- a/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/blas/CpuLevel3.java
+++ b/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/blas/CpuLevel3.java
@@ -17,11 +17,15 @@
 package org.nd4j.linalg.cpu.nativecpu.blas;
 
 
+import lombok.val;
 import org.bytedeco.javacpp.DoublePointer;
 import org.bytedeco.javacpp.FloatPointer;
 import org.nd4j.linalg.api.blas.impl.BaseLevel3;
+import org.nd4j.linalg.api.blas.params.MMulTranspose;
+import org.nd4j.linalg.api.buffer.DataType;
 import org.nd4j.linalg.api.ndarray.INDArray;
 import org.nd4j.linalg.api.ops.aggregates.impl.AggregateGEMM;
+import org.nd4j.linalg.api.ops.impl.reduce.Mmul;
 import org.nd4j.linalg.factory.Nd4j;
 import org.nd4j.nativeblas.Nd4jBlas;
 
@@ -41,7 +45,25 @@ public class CpuLevel3 extends BaseLevel3 {
     @Override
     protected void hgemm(char Order, char TransA, char TransB, int M, int N, int K, float alpha, INDArray A, int lda,
                     INDArray B, int ldb, float beta, INDArray C, int ldc) {
-        throw new UnsupportedOperationException();
+/*
+        if (false) {
+            val fA = A.castTo(DataType.FLOAT);
+            val fB = B.castTo(DataType.FLOAT);
+            val fC = Nd4j.create(DataType.FLOAT, C.shape(), C.ordering());
+
+            sgemm(Order, TransA, TransB, M, N, K, alpha, fA, lda, fB, ldb, beta, fC, ldc);
+
+            C.assign(fC);
+        } else {
+            */
+            val t = MMulTranspose.builder()
+                    .transposeA(TransA == 'T')
+                    .transposeB(TransB == 'T')
+                    .transposeResult(false)
+                    .build();
+            val op = new Mmul(A, B, C, t);
+            Nd4j.exec(op);
+        //}
     }
 
     @Override

--- a/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/blas/CpuLevel3.java
+++ b/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/blas/CpuLevel3.java
@@ -45,25 +45,26 @@ public class CpuLevel3 extends BaseLevel3 {
     @Override
     protected void hgemm(char Order, char TransA, char TransB, int M, int N, int K, float alpha, INDArray A, int lda,
                     INDArray B, int ldb, float beta, INDArray C, int ldc) {
-/*
-        if (false) {
+
+        //if (true) {
             val fA = A.castTo(DataType.FLOAT);
             val fB = B.castTo(DataType.FLOAT);
-            val fC = Nd4j.create(DataType.FLOAT, C.shape(), C.ordering());
+            val fC = C.castTo(DataType.FLOAT);
 
             sgemm(Order, TransA, TransB, M, N, K, alpha, fA, lda, fB, ldb, beta, fC, ldc);
 
             C.assign(fC);
-        } else {
-            */
+        /*} else {
+            // TODO: uncomment this once we have optimized gemm calls
             val t = MMulTranspose.builder()
-                    .transposeA(TransA == 'T')
-                    .transposeB(TransB == 'T')
+                    .transposeA(false)
+                    .transposeB(false)
                     .transposeResult(false)
                     .build();
             val op = new Mmul(A, B, C, t);
             Nd4j.exec(op);
-        //}
+        }
+         */
     }
 
     @Override

--- a/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/blas/BlasTests.java
+++ b/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/blas/BlasTests.java
@@ -18,6 +18,8 @@ package org.nd4j.linalg.blas;
 
 
 import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
@@ -26,6 +28,9 @@ import org.nd4j.linalg.api.buffer.DataType;
 import org.nd4j.linalg.api.ndarray.INDArray;
 import org.nd4j.linalg.factory.Nd4j;
 import org.nd4j.linalg.factory.Nd4jBackend;
+
+import java.util.ArrayList;
+import java.util.Collections;
 
 import static org.junit.Assert.*;
 
@@ -198,7 +203,7 @@ public class BlasTests extends BaseNd4jTest {
     }
 
     @Test
-    public void test_Fp16_Mmuli1(){
+    public void test_Fp16_Mmuli_1(){
         final INDArray activations = Nd4j.createUninitialized(DataType.HALF, new long[]{1, 3, 2}, 'f');
         final INDArray z = activations.tensorAlongDimension(0, 1, 2);
 
@@ -208,6 +213,35 @@ public class BlasTests extends BaseNd4jTest {
         INDArray ab = a.mmul(b);
         a.mmul(b, z);
         assertEquals(ab, z);
+    }
+
+    @Test
+    public void test_Fp16_Mmuli_2(){
+        val a = Nd4j.create(DataType.HALF, 32, 768);
+        val b = Nd4j.create(DataType.HALF, 768);
+
+        val c = a.mmul(b);
+    }
+
+    @Test
+    @Ignore
+    public void testHalfPrecision() {
+        val a = Nd4j.create(DataType.HALF, 64, 768);
+        val b = Nd4j.create(DataType.HALF, 768, 1024);
+        val c = Nd4j.create(DataType.HALF, new long[]{64, 1024}, 'f');
+
+        val durations = new ArrayList<Long>();
+        val iterations = 100;
+        for (int e = 0; e < iterations; e++) {
+            val timeStart = System.currentTimeMillis();
+            a.mmuli(b, c);
+            val timeEnd = System.currentTimeMillis();
+            durations.add(timeEnd - timeStart);
+        }
+
+        Collections.sort(durations);
+
+        log.info("Median time: {} ms", durations.get(durations.size() / 2));
     }
 
     @Test

--- a/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/blas/BlasTests.java
+++ b/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/blas/BlasTests.java
@@ -198,6 +198,19 @@ public class BlasTests extends BaseNd4jTest {
     }
 
     @Test
+    public void test_Fp16_Mmuli1(){
+        final INDArray activations = Nd4j.createUninitialized(DataType.HALF, new long[]{1, 3, 2}, 'f');
+        final INDArray z = activations.tensorAlongDimension(0, 1, 2);
+
+        final INDArray a = Nd4j.rand(DataType.HALF, 3, 4);
+        final INDArray b = Nd4j.rand(DataType.HALF,4, 2);
+
+        INDArray ab = a.mmul(b);
+        a.mmul(b, z);
+        assertEquals(ab, z);
+    }
+
+    @Test
     public void testMmuli4(){
         try {
             Nd4j.rand(1, 3).mmuli(Nd4j.rand(3, 1), Nd4j.createUninitialized(new int[]{10, 10, 1}));

